craft_parts/parts.py

```diff
@@ -205,8 +206,16 @@ def part_build_dir(self) -> Path:
 
     @property
     def part_build_subdir(self) -> Path:
-        """Return the subdirectory in build containing the source subtree (if any)."""
-        if self.spec.source_subdir:
+        """Return the subdirectory in build containing the source subtree (if any).
+
+        Parts that have a source subdirectory and do not support out-of-source builds
+        will have a build subdirectory.
+        """
+        if (
+            self.plugin_name != ""
+            and self.spec.source_subdir
+            and not get_plugin_class(self.plugin_name).get_out_of_source_build()
```

Comment:
This is using a new `@classmethod` called `get_out_of_source_build()`.

I couldn't use the existing `out_of_source_build` property, because I would have to create an instance of the plugin.  That gets tricky and circular, as the plugin needs part info, but this code is in the part constructor.